float32 desx
float32 desy
float32 desz
float32 desx_p
float32 desy_p
float32 desz_p
float32 despsi
float32 despsi_p